src/cargo/ops/tree/graph.rs

```diff
@@ -0,0 +1,562 @@
+//! Code for building the graph used by `cargo tree`.
+
+use super::TreeOptions;
+use crate::core::compiler::{CompileKind, RustcTargetData};
+use crate::core::dependency::DepKind;
+use crate::core::resolver::features::{FeaturesFor, RequestedFeatures, ResolvedFeatures};
+use crate::core::resolver::Resolve;
+use crate::core::{
+    FeatureMap, FeatureValue, InternedString, Package, PackageId, PackageIdSpec, Workspace,
+};
+use crate::util::CargoResult;
+use std::collections::{HashMap, HashSet};
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub enum Node {
+    Package {
+        package_id: PackageId,
+        /// Features that are enabled on this package.
+        features: Vec<InternedString>,
+        kind: CompileKind,
+    },
+    Feature {
+        /// Index of the package node this feature is for.
+        node_index: usize,
+        /// Name of the feature.
+        name: InternedString,
+    },
+}
+
+#[derive(Debug, Copy, Hash, Eq, Clone, PartialEq)]
+pub enum Edge {
+    Dep(DepKind),
+    Feature,
+}
+
+#[derive(Clone)]
+struct Edges(HashMap<Edge, Vec<usize>>);
```

Comment:
In addition to the comments below on `edges`, I'm a little lost here (again I haven't read the whole module, so this is a comment from just reading the data structures) about why there's a `Vec<HashMap<_, Vec>>` here, so mind adding a comment here explaining a bit?